contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,239 @@ The central manager will poll each worker for their discovered volumes and creat
+* list/get/update StorageClasses
+
+##### Volume Creation
+
+During start-up of the provisioner, a new work queue `ProvisionQueue` will be created,
+and passed into both `populator` and `storageManager`:
+
+```
+ProvisionQueue *workqueue.Type
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will pass it into the queue to trigger volume privisioning;
+`StorageManager` will be the consumer, it will watch the queue and
+call its `CreateLocalVolume` function to:
+
+1. Create volumes behind a PV accordingly
+2. Create a PV object that is pre-bound to the PVC
+
+The proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if `annProvisionerTopology` is set on the PVC. If not, then it means
+that the scheduler made a provisioning decision before the capacity initialized
+in StorageClasses. Return error.
+3. Check if the class is available for dynamic provisioning
+(e.g. has an input of `volumeGroup` for LVM). If not, return error.
+4. Create volumes behind a PV (e.g. for LVM, Create `logical volume (LV)`
+object in the VG).
+5. Create a PV API object that is pre-bound to the PVC.
+
+**Note:** As it describes in [Volume Topology-aware Scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+in whichever step it fails, the provisioner will signal back to the scheduler
+to retry dynamic provisioning by removing the "annSelectedNode" annotation of
+the PVC.
+
+For the purpose, it requires the permission to update PersistentVolumeClaims.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+Existing script `quick_reset.sh` calls `mkfs -F` and `wipefs` to cleanup a block.
+In addition to that, a hook named `volumeDeleteFunc` will be added into `deleter`,
+and for dynamic privisioned volumes, the deleting process should be:
+1. Call scripts/binaries in `blockCleanerCommand` to clean up data in the volume.
+2. Call `volumeDeleteFunc` to recycle the volume behind the PV (e.g. delete the LV for lvm).
+3. Delete the PV object.
+
+##### Expansion of Local Volume Plugin
+
+Currently, Intended `VolumeMode` of a local volume is determined by the state of
+the input path. If a user need a local volume whose `VolumeMode` is filesystem,
+the path provided in `LocalVolumeSource` must be a directory. Thus, for dynamic
+provisioned volumes, the provisioner need to mount the LV to a global path first,
+to allow kubelet for another mount. It increases the complexity of the plugin.
+
+Therefore, in-tree local storage plugin will be updated to support both directory
+and block as volume source.

```

Comment:
When the specified Path is detected to be a block device and the volumeMode is filesystem, the plugin will create the filesystem (on the first mount) and globally mount the device to the plugin path.  This is similar to how other volume plugins handle filesystem on block.